core/controllers/admin.py

```diff
@@ -741,37 +763,125 @@ def get(self):
             if user_id is None:
                 raise self.InvalidInputException(
                     'User with given username does not exist.')
-            user_role_dict = {
-                username: user_services.get_user_role_from_id(user_id)
+
+            user_settings = user_services.get_user_settings(user_id)
+            user_roles = user_settings.roles
+            topic_ids = []
+            if feconf.ROLE_ID_TOPIC_MANAGER in user_roles:
+                topic_ids = [
+                    t_r.id for t_r in topic_fetchers.get_topic_rights_with_user(
+                        user_id)]
+            user_roles_dict = {
+                'roles': user_roles,
+                'topic_ids': topic_ids,
+                'banned': user_settings.banned
             }
-            self.render_json(user_role_dict)
+            self.render_json(user_roles_dict)
 
     @acl_decorators.can_access_admin_page
-    def post(self):
-        username = self.normalized_payload.get('username')
-        role = self.normalized_payload.get('role')
-        topic_id = self.normalized_payload.get('topic_id')
+    def put(self):
+        username = self.payload.get('username')
+        role = self.payload.get('role')
+        topic_id = self.payload.get('topic_id')
+        user_settings = user_services.get_user_settings_from_username(username)
+
+        if user_settings is None:
+            raise self.InvalidInputException(
+                'User with given username does not exist.')
+
+        if role == feconf.ROLE_ID_TOPIC_MANAGER:
+            if feconf.ROLE_ID_TOPIC_MANAGER not in user_settings.roles:
+                user_services.add_user_role(user_settings.user_id, role)
+
+            user = user_services.get_user_actions_info(user_settings.user_id)
+            topic_services.assign_role(
+                user_services.get_system_user(), user,
+                topic_domain.ROLE_MANAGER, topic_id)
+        else:
+            user_services.add_user_role(user_settings.user_id, role)
+
+        self.render_json({})
+
+    @acl_decorators.can_access_admin_page
+    def delete(self):
+        username = self.request.get('username')
+        role = self.request.get('role')
+        topic_id = self.request.get('topic_id')
+        remove_from_all_topics = self.request.get(
+            'remove_from_all_topics', False)
+
         user_id = user_services.get_user_id_from_username(username)
         if user_id is None:
             raise self.InvalidInputException(
                 'User with given username does not exist.')
 
-        if (
-                user_services.get_user_role_from_id(user_id) ==
-                feconf.ROLE_ID_TOPIC_MANAGER):
+        if role != feconf.ROLE_ID_TOPIC_MANAGER:
+            user_services.remove_user_role(user_id, role)
+            self.render_json({})
+            return
+
+        if remove_from_all_topics:
```

Comment:
> Why do we need to pass a separate bool and rely on the client for this? I think we should check the topic list after doing any operations and remove the user role then if needed.

We are doing the check internally and we are not relying on client. But in case client wants to remove the TOPIC_MANAGER role completely from the user we are sending remove_from_all_topics=True.

Current strucutre:
```
To remove from one topic (can remove topic_manager role if user only manages 1 topic):
role: TOPIC_MANGER
topic_id: <topicId>

To remove from all topics and remove topic_manager role:
role: TOPIC_MANGER
topic_id: <topicId>
```
Another possible solution:
```
To remove from one topic (can remove topic_manager role if user only manages 1 topic):
remove_from_topic_id: <topicId>

To remove from all topics and remove topic_manager role:
role: TOPIC_MANGER
```

Let me know whether the "another" solution sounds okay? (I'm trying to make it either remove 1 or all, as we are handling one role editing at a time through this handler.)

Reply:
I would say that the cleanest option is: for any operations related to the topic manager role, they must **always** include a topic ID. The end user never manages that role directly.

You might even go so far as to have an entirely separate handler for operations related to topic management.

Reply:
Create seperate handler for topic assignment/deassignment.